### ISSUE TITLE
p-dropdown needs double click to select

### ISSUE THREAD

**Issue by i-amn** [primefaces/primeng#6913]
Current behavior
While using primeng dropdown needs double click to select the value. Pushing Enter key works fine.

Angular version: 7

PrimeNG version: 6.1.5

Browser: [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

Language: [all | TypeScript X.X | ES6/7 | ES5]

Hi,

This is still an issue.

https://stackblitz.com/edit/test-p-table-msxkby?file=package.json

I just updated the stackblitz to primeng 6.
https://stackoverflow.com/questions/51720979/primeng-dropdown-needs-doubleclick-when-using-a-function-for-options

Thank you,
Aman

**Comment by i-amn** [comment]
This needs to be resolved, what if we want to show options in for loop and according to the key?

**Comment by Mysame** [comment]
This has nothing to do with PrimeNG.

You keep re-creating the options.
Angular keeps calling `getOptions2()` every update, which results in you clicking an object that either no longer exists, or has been replaced.

Bind options to a list that only updates when required, *not* one that gets re-created every Angular update.

**Comment by i-amn** [comment]
@Mysame i am not able to understand you said 

> 'Angular keeps calling getOptions2() every update, ' that's what we want to create options on every update.

for now can you suggest how p-autocomplete working fine with the same code? :)

**Comment by Mysame** [comment]
You can find dozens of posts about [change detection](https://blog.thoughtram.io/angular/2016/02/22/angular-2-change-detection-explained.html).
Simply put, every time _anything_ changes in your app, Angular runs change detection. This will re-validate and re-calculate all bindings. Including your code of `[options]="getOptions2()`. Every time this happens, `getOptions2()` will return a brand new list. Data might go out of sync, where the GUI is showing old data, but the binding is new. Hence, clicking it doesn't do anything. But clicking triggers change detection _again_, which will repopulate the dropdown.

Long story short: **don't** put functions as bindings.

https://stackblitz.com/edit/test-p-table-rsjpqd

**Comment by i-amn** [comment]
@Mysame and again i am asking same question thrice 

> "for now can you suggest how it is working when we are double clicking on it?"

**Comment by Mysame** [comment]
> with the same code

You don't, the code is wrong.
I'm unsubscribing from this issue, as it is clearly a waste of my time

**Comment by cagataycivici** [comment]
Not PrimeNG related, avoid recreating options at the binding.
